### ISSUE TITLE
How use GRU in opacus?

### ISSUE THREAD

**Issue by anik123** [pytorch/opacus#205]
I have tried to use Gated Recurrent Unit (GRU) with opacus. However, getting `IncompatibleModuleException: Model contains incompatible modules.`.

Is there any way to use GRU with opacus? 

Thanks for your time.

**Comment by ashkan-software** [comment]
Hi @anik123 

Opacus by default does not support GRU. By looking at the definition of Wikipedia, I assume it's because of the forget gate, but I am not sure :).
 
See the list of supported modules here:

https://github.com/pytorch/opacus/tree/main/opacus#supported-modules 

There are two ways that Opacus can support modules. A module can be either a _building block_, or a _composite_:

_Building block_. These are “atomic” trainable modules (i.e., default classes) that have their own hooks, and can be used directly, namely, `nn.Linear`, `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`, and the normalization layers (`nn.LayerNorm`, `nn.GroupNorm`, `nn.InstanceNorm`).

_Composite_. These are modules that are composed of building blocks and they are supported as long as all trainable submodules are supported (frozen submodules need not be supported.) We made a DP interface to replace composite modules we do not support, such as `nn.MultiheadAttention` with re-implementation that use only supported building blocks. 

Needless to say, Opacus supports modules with no trainable parameters (e.g. `nn.ReLU`, and `nn.Tan`h) and modules that are frozen. These modules are supported out of the box, as there is no need for computing gradients for these. 

https://opacus.ai/docs/faq#my-model-throws-incompatiblemoduleexception-what-is-going-wrong 

I am closing this task, but if you think your question is not answered, feel free to reopen this task! :)

**Comment by anik123** [comment]
@ashkan-software Thanks for your answer. Why opacus is so restricted? Is there any plan to provide GRU support? like DPLSTM?

**Comment by ashkan-software** [comment]
Hi @anik123 

Please see the feature request I just created here: https://github.com/pytorch/opacus/issues/206

Currently, we do not plan to work on this feature. However, we encourage contributions from the open-source community. ☺️

**Comment by anik123** [comment]
Hi @ashkan-software ,

Thanks for your kind consideration.
